modules/fundamental/src/analysis/service/mod.rs

```diff
@@ -0,0 +1,584 @@
+            ancestors,
+        }
+    }
+}
+
+pub struct AnalysisService {
+    db: Database,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn add_node(
+    g: &mut Graph<PackageNode, Relationship, petgraph::Directed>,
+    nodes: &mut HashMap<String, NodeIndex>,
+    component_purl: &String,
+    component_name: &String,
+    sbom_published: &String,
+    document_id: &String,
+    product_name: &String,
+    product_version: &String,
+) -> NodeIndex {
+    match nodes.get(&component_purl.to_string()) {
+        Some(&i) => i,
+        None => {
+            let i = g.add_node(PackageNode {
+                purl: component_purl.to_string(),
+                name: component_name.to_string(),
+                published: sbom_published.to_string(),
+                document_id: document_id.to_string(),
+                product_name: product_name.to_string(),
+                product_version: product_version.to_string(),
+            });
+            nodes.insert(component_purl.to_string(), i);
+            i
+        }
+    }
+}
+
+pub fn regex_purl_find(nodes: &HashMap<String, NodeIndex>, search_value: &str) -> Vec<NodeIndex> {
+    // Sanitize the search value to prevent regex injection attacks
+    let sanitized_search_value = regex::escape(search_value);
+
+    let pattern = match Regex::new(&sanitized_search_value) {
+        Ok(pattern) => pattern,
+        Err(_) => return Vec::new(), // Return an empty vector if the regex pattern is invalid
+    };
+
+    nodes
+        .iter()
+        .filter(|(key, _)| pattern.is_match(key))
+        .map(|(_, value)| *value)
+        .collect()
+}
+
+pub fn exact_name_find(nodes: &HashMap<String, NodeIndex>, search_value: &str) -> Vec<NodeIndex> {
+    nodes
+        .iter()
+        .filter_map(|(key, &value)| {
+            Purl::from_str(key).ok().and_then(|purl| {
+                if purl.name == search_value {
+                    Some(value)
+                } else {
+                    None
+                }
+            })
+        })
+        .collect()
+}
+
+pub fn ancestor_nodes(
+    graph: &petgraph::Graph<PackageNode, Relationship, petgraph::Directed>,
+    node: NodeIndex,
+    original_node: NodeIndex,
+) -> Vec<NodeIndex> {
+    // we need order
+    let mut ancestor_nodes: Vec<NodeIndex> = Vec::new();
+    if node != original_node {
+        ancestor_nodes.push(node);
+    }
+    fn dfs(
+        graph: &petgraph::Graph<PackageNode, Relationship, petgraph::Directed>,
+        node: NodeIndex,
+        original_node: NodeIndex,
+        ancestor_nodes: &mut Vec<NodeIndex>,
+    ) {
+        for neighbor in graph.neighbors_directed(node, Direction::Outgoing) {
+            if neighbor == original_node {
+                break; // no need to descend further past the original node in the tree
+            }
+            if !ancestor_nodes.contains(&neighbor) {
+                ancestor_nodes.push(neighbor);
+                dfs(graph, neighbor, original_node, ancestor_nodes);
```

Comment:
Are we concerned about recursion depth at all? Worth translating to a stack/queue-based algorithm for depth-first?

Reply:
Yea, there should also be some checks/tests for loops in SBOMs. Which might be illegal, but also shouldn't crash the system.